pyiron_contrib/tinybase/murn.py

```diff
@@ -0,0 +1,94 @@
+from pyiron_contrib.tinybase.container import (
+            AbstractOutput,
+            StructureInput,
+)
+from pyiron_contrib.tinybase.node import (
+            AbstractNode,
+            ListNode,
+            ListInput,
+            ReturnStatus
+)
+
+from copy import deepcopy
+
+import numpy as np
+import matplotlib.pyplot as plt
+import scipy.interpolate as si
+import scipy.optimize as so
+
+from pyiron_atomistics.atomistics.structure.has_structure import HasStructure
+
+MurnaghanInputBase = StructureInput.from_attributes(
+        "MurnaghanInputBase",
+        "strains",
+        "node"
+)
+
+class MurnaghanInput(MurnaghanInputBase, ListInput):
+    def check_ready(self):
+        structure_ready = self.structure is not None
+        strain_ready = len(self.strains) > 0
+        node = self.node
+        node.input.structure = self.structure
+        return structure_ready and strain_ready and node.input.check_ready()
+
+    def set_strain_range(self, range, steps):
+        self.strains = (1 + np.linspace(-range, range, steps))**(1/3)
+
+    def _create_nodes(self):
+        cell = self.structure.get_cell()
+        nodes = []
+        for s in self.strains:
+            n = deepcopy(self.node)
+            n.input.structure = self.structure.copy()
+            n.input.structure.set_cell(cell * s, scale_atoms=True)
+            nodes.append(n)
+        return nodes
+
+MurnaghanOutputBase = AbstractOutput.from_attributes(
+        "MurnaghanOutputBase",
+        'base_structure',
+        volumes=list,
+        energies=list
+)
+
+class MurnaghanOutput(MurnaghanOutputBase, HasStructure):
+    def plot(self):
+        plt.plot(self.volumes, self.energies)
+
+    @property
+    def equilibrium_volume(self):
+        inter = si.interp1d(self.volumes, self.energies)
+        return so.minimize_scalar(inter, bounds=(np.min(self.volumes), np.max(self.volumes))).x
+
+    def _number_of_structures(self):
+        return 1
+
+    def _get_structure(self, frame, wrap_atoms=True):
+        s = self.base_structure
+        s.set_cell(s.get_cell() * (self.equilibrium_volume/s.get_volume())**(1/3))
+        return s
+
+class MurnaghanNode(ListNode):
+
+    def _get_input(self):
+        return MurnaghanInput()
+
+    def _get_output(self):
+        return MurnaghanOutput()
+
+    def _execute(self, output):
```

Comment:
Why does `_execute` take _output_ as an argument?? Something has gone sideways with the naming scheme here.

Reply:
> which is what everyone has to implement

So I think one of the things I'm running up against here is the volume of moving pieces when defining a new node. If I understand correctly, first new input and output classes need to be defined (inheriting from their respective `Abstract`s) -- what I like about this is the ability to leverage composition. 

Then the user defines `_execute`, which is the real meat of the node, this takes an instance of the output as an argument; internally the user is then expected to leverage the attributes of `self.input` to modify the attributes of `output`. The `execute` wrapper for `_execute` then makes sure the modified output instance gets returned. 

The main thing I don't like about this is how implicit everything is when defining the key aspect of node behaviour -- when defining `_execute` the dev needs to refer back to their IO classes all the time; they might be able to get decent variable name hints for `output` if they make sure to type hint `_execute(self, output: MyOutputClass)`, but they will never be able to get good hints for `self.input` from their IDE because it is (perfectly correctly) type hinted at the parent class level as `AbstractInput`. 

Additionally it looks like the output class...can have default values? `AbstractOutput.from_attributes(foo=list)` uses the `list` constructor to make an empty one? Then I see big danger for a dev making their output class, forgetting to populate parts of it, and silently getting totally weird returned values.

> but if _execute produces the output itself instead of getting it passed, then `execute` would have to do a type check afterwards

I don't think I agree with that; why would we type check outputs? If I write `def foo(x: int) -> str: return x`, that's _my_ fault! 😜 In the same way, if they say they're returning an instance `MyMurnOutput` and return `MyFooBarOutput`, I don't see the need for us to catch them.

Reply:
> The main thing I don't like about this is how implicit everything is when defining the key aspect of node behaviour -- when defining `_execute` the dev needs to refer back to their IO classes all the time; they might be able to get decent variable name hints for `output` if they make sure to type hint `_execute(self, output: MyOutputClass)`, but they will never be able to get good hints for `self.input` from their IDE because it is (perfectly correctly) type hinted at the parent class level as `AbstractInput`.

I'm not sure I see the problem.  Is the need the refer back to their input/output not part of any solution that we'd come up with?  If you define the node input implicitly based on the function arguments they'd have everything visible in their definition of `_execute`, but it'd be very easy to forget some parts of the arguments of all the input they ought to support to be general, so they still need to check if the function that they wrote actually match the generic interface by... looking at the corresponding generic input class.

As for type hinting of input, I think we can come up with something to make that work, e.g. node devs annotating their `_get_input` with the input type.



> Additionally it looks like the output class...can have default values? `AbstractOutput.from_attributes(foo=list)` uses the `list` constructor to make an empty one? Then I see big danger for a dev making their output class, forgetting to populate parts of it, and silently getting totally weird returned values.

Fair, but imo solvable by working on the `from_attributes` or `StorageAttribute` builders.



> > but if _execute produces the output itself instead of getting it passed, then `execute` would have to do a type check afterwards
> 
> I don't think I agree with that; why would we type check outputs? If I write `def foo(x: int) -> str: return x`, that's _my_ fault! 😜 In the same way, if they say they're returning an instance `MyMurnOutput` and return `MyFooBarOutput`, I don't see the need for us to catch them.

Ja, here I very strongly disagree.  For things that are defined in the notebook interactively you are right that it doesn't matter much, but for "official" nodes that live really in one of the pyiron modules I'd like to give really strong guarantees that they work as advertised.  Otherwise the same problems that currently haunt the generic methods like `calc_*` in atomistics will come back, i.e. you explicitly need to check that sub classes take the same arguments with the same semantics.